File: src/main/java/net/dashflight/data/postgres/FlywayMigrator.java
package net.dashflight.data.postgres;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.util.Map;

/**
 * Migrates the test container to the correct version and returns a client connected to it.
 */
public class FlywayMigrator {

    public static void migrateAndExecute(PostgresContainer container, PostgresTestCase testCase) throws Exception {
        if (container == null) {
            throw new IllegalArgumentException("Container cannot be null");
        }

        String address = container.getContainerIpAddress();
        int port = container.getFirstMappedPort();

        Map<String, String> env = container.getEnvMap();

        String username = env.get("POSTGRES_USER");
        String password = env.get("POSTGRES_PASSWORD");
        String database = env.get("POSTGRES_DB");


        PostgresConnectionOptions options = PostgresConnectionOptions.builder()
                .host(address)
                .port(port)
                .username(username)
                .password(password)
                .dbname(database)
                .build();


        try (PostgresClient pgClient = new PostgresClient(() -> options)) {
            try {
                new FlywayManager(pgClient).getFlyway().migrate();
            } catch (IOException | NoSuchAlgorithmException e) {
                e.printStackTrace();
                throw new RuntimeException("Flyway could not migrate database");
            }


            testCase.run(pgClient);
        }
    }


    interface PostgresTestCase {
        void run(PostgresClient client) throws Exception;
    }
}

File: src/main/java/net/dashflight/data/postgres/FlywayMigrator.kt
package net.dashflight.data.postgres

import java.io.IOException
import java.security.NoSuchAlgorithmException

/**
 * Migrates the test container to the correct version and returns a client connected to it.
 */
object FlywayMigrator {

    @Throws(Exception::class)
    fun migrateAndExecute(container: PostgresContainer, testCase: (client: PostgresClient) -> Unit) {
        val address = container.containerIpAddress
        val port = container.firstMappedPort
        val env = container.envMap

        val username = env["POSTGRES_USER"]
        val password = env["POSTGRES_PASSWORD"]
        val database = env["POSTGRES_DB"]

        val options = PostgresConnectionOptions(address, port, database, username, password)

        PostgresClient(object : PostgresConnectionOptionProvider {
            override fun get(): PostgresConnectionOptions {
                return options
            }
        }).use { pgClient ->
            try {
                FlywayManager(pgClient).flyway.migrate()
            } catch (e: IOException) {
                e.printStackTrace()
                throw RuntimeException("Flyway could not migrate database")
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
                throw RuntimeException("Flyway could not migrate database")
            }
            testCase(pgClient)
        }
    }
}